infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesClientFactory.java

```diff
@@ -43,6 +53,59 @@ public KubernetesClientFactory(
       @Nullable @Named("che.infra.kubernetes.password") String password,
       @Nullable @Named("che.infra.kubernetes.oauth_token") String oauthToken,
       @Nullable @Named("che.infra.kubernetes.trust_certs") Boolean doTrustCerts) {
+    this.defaultConfig =
+        buildDefaultConfig(masterUrl, username, password, oauthToken, doTrustCerts);
+    this.httpClient = HttpClientUtils.createHttpClient(defaultConfig);
+  }
+
+  /**
+   * Builds the Kubernetes {@link Config} object based on a default {@link Config} object and an
+   * optional workspace Id.
+   */
+  public Config buildConfig(Config defaultConfig, @Nullable String workspaceId)
+      throws InfrastructureException {
+    return defaultConfig;
+  }
+
+  /**
+   * Creates instance of {@link KubernetesClient}.
+   *
+   * @throws InfrastructureException if any error occurs on client instance creation.
+   */
+  public KubernetesClient create(String workspaceId) throws InfrastructureException {
+    Config configForWorkspace = buildConfig(defaultConfig, workspaceId);
+
+    return create(configForWorkspace);
+  }
+
+  /**
+   * Creates instance of {@link KubernetesClient}.
+   *
+   * @throws InfrastructureException if any error occurs on client instance creation.
+   */
+  public KubernetesClient create() throws InfrastructureException {
```

Comment:
I think this method should be removed at all. It may lead to an error because default client is not supposed to be used for performing an operation with OS cluster, while it can be used.
As far as I see it is used only for parsing Kubernetes list. Then it makes sense to create default Kubernetes client there or use io.fabric8.kubernetes.client.utils.Serialization#unmarshal method directly.

Reply:
> I think this method should be removed at all. It may lead to an error because default client is not supposed to be used for performing an operation with OS cluster, while it can be used.

I don't understand the problem or the error this method could lead to: it's still calling the default configuration, it's still calling `buildConfig()` method, though without a workspace ID.

Reply:
Functionally it's the exactly the same method as the previous `create()` method.
Once again, maybe we can remove it and change the `list()` method to something else, but I'm not sure it has its place in this PR.

Reply:
> Functionally it's the exactly the same method as the previous create() method.

Yeah, but previously there was only one method.
As far as I understand, Now clients **must** be created with specified `workspaceId`.

> I don't understand the problem or the error this method could lead to: it's still calling the default configuration, it's still calling buildConfig() method, though without a workspace ID.

Within Che it will work even if all components of Kubernetes/OpenShift infrastructure will use default client without specifying`workspaceId`. But will RhChe work in this case? You'll update all places to create clients with workspaceId. But it's not clear enough where workspace id must be specified and where it is optional. 
So new components may create default clients and I guess it will lead to failing RHChe packaging. WDYT? Does it make sense?

Reply:
> Now clients *must* be created with specified workspaceId

Well, not exactly. That's more: *when the `workspaceId` is known then it __should__ be provided* to help inferring the right OpenShift cluster / namespace to connect to.

However if no `workspaceId` is available then `rh-che` would simply use the current user information to infer the right Openshift cluster / namespace. I don't see the problem here: it's mainly the same behavior as on Che 5, but implemented in a cleaner way IMO.

Reply:
> Now clients must be created with specified workspaceId

Oh, It was my wrong assumption. It's quite confusing. Maybe a good java doc will make it more clear.

But is default client really needed? If no, please consider removing default client at all to make it more straightforward. As far as I understand `workspaceId` is unknown only if this operation is not really related to a particular workspace. But then why should we interact with Kubernetes/OpenShift API? It could be a case like fetch cluster information, quotas or something like that. But I think it'd better remove such method while we don't need them. So, I don't have strong arguments to do it, so **up to you** remove it or not

Reply:
@davidfestal Please add java docs that make clear when a client should be created with specified workspace id and when default one without parameters.

Reply:
Done in commit https://github.com/eclipse/che/pull/8805/commits/5ae5218d17217e07557baa62f6523b3964262f8a